### PR TITLE
Allow configuring mplex

### PR DESCRIPTION
**Pull request by tomaka**
This allows configuring the behaviour of mplex.

Right now, if the internal buffer is full we produce an error on all substreams.
This is the default behaviour because it guarantees that we can't have any deadlock.

However there are situations, such as in polkadot, where we know that all substreams are independant (ie. there's no situation processing of one substream doesn't block the processing of another substream).
This PR allows configuring this behaviour.

### CODE REVIEW

File: mplex/src/lib.rs [hunk 1]
```diff
@@ -44,21 +44,74 @@ use futures::{future, stream::Fuse, task};
...
+        }
+    }
+}
+
+/// Behaviour when the maximum length of the buffer is reached.
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub enum MaxBufferBehaviour {
+    /// Produce an error on all the substreams.
+    CloseAll,
+    /// No new message will be read fro mthe underlying ocnnection if the buffer is full.
```
Comment (gnunicorn) [inline mplex/src/lib.rs]: typo "fro**m** the"

File: mplex/src/lib.rs [hunk 2]
```diff
@@ -138,18 +197,45 @@ where C: AsyncRead + AsyncWrite,
...
 
+    // TODO: replace with another system
+    static NEXT_TASK_ID: AtomicUsize = AtomicUsize::new(0);
+    task_local!{
+        static TASK_ID: usize = NEXT_TASK_ID.fetch_add(1, Ordering::Relaxed)
+    }
+
     loop {
+        // Check if we reached max buffer length first.
+        if inner.buffer.len() == inner.config.max_buffer_len {
```
Comment (gnunicorn) [inline mplex/src/lib.rs]: Why not a `>=`? Can we be sure to always reach the `max_buffer_len` exactly?

Comment (tomaka) [reply mplex/src/lib.rs]: I'll add a `debug_assert!` instead, so that we detect logic errors in the code.

Comment (gnunicorn) [reply mplex/src/lib.rs]: if you think that satisfied it. We had a `>=` before...

**Pull request merged**